src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -2467,41 +2561,28 @@ public override BoundNode VisitMethodGroup(BoundMethodGroup node)
             return null;
         }
 
-        /// <summary>
-        /// Unconditionally visits an expression.
-        /// If the expression has "state when not null" after visiting,
-        /// the method returns 'true' and writes the state to <paramref name="stateWhenNotNull" />.
-        /// </summary>
-        protected bool VisitPossibleConditionalAccess(BoundExpression node, [NotNullWhen(true)] out TLocalState? stateWhenNotNull)
+        private bool TryVisitConditionalAccess(BoundExpression node, [NotNullWhen(true)] out TLocalState? stateWhenNotNull)
```

Comment:
>TryVisitConditionalAccess [](start = 21, length = 25)

nit: doc comment for contract would be nice (if node is a conditional access, visit it and return true, false otherwise)

Reply:
It occurs to me that there is a possibility to double-visit nodes using this method.

e.g. `true == ((C)null)?.M(x = 1)` doesn't have conditional state after visiting. But we may return 'false' from this method and visit the operand again.

Reply:
I think the scenarios where this may happen are rare and simple enough that it isn't worth modifying to try and prevent the double visit.

Will add the doc comment for contract as suggested.

Reply:
> But we may return 'false' from this method and visit the operand again.

In terms of expected behavior, I think we should only visit once. I don't know if we have a hard rule on that.
In terms of actual behavior, I couldn't see how we might visit twice.

---
In reply to: [610049290](https://github.com/dotnet/roslyn/pull/52425#discussion_r610049290) [](ancestors = 610049290)

Reply:
Synced offline, have pushed a change to ensure we don't visit multiple times. (it's an edge case, but still helps it feel simpler.)